### PR TITLE
Move core files to src folder and move JS files to TypeScript

### PR DESCRIPTION
**Pull request by timneutkens**
- Move core files to `src` folder at `packages/next/src`.
- Refactor all `.js` files to TypeScript, note that many of these were external packages so I've added `any` for many of them but it caught a few bugs already. Just making sure all code is consistently TypeScript.

First in a series of smaller changes to the folder structure.

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have a helpful link attached, see [`contributing.md`](https://github.com/vercel/next.js/blob/canary/contributing.md)

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] [e2e](https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs) tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have a helpful link attached, see [`contributing.md`](https://github.com/vercel/next.js/blob/canary/contributing.md)

## Documentation / Examples

- [ ] Make sure the linting passes by running `pnpm build && pnpm lint`
- [ ] The "examples guidelines" are followed from [our contributing doc](https://github.com/vercel/next.js/blob/canary/contributing/examples/adding-examples.md)

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: bf492f418ec76c532f87fada3bd2eec3e74b39ef

`pnpm testheadless test/unit/web-runtime/next-response.test.ts` 

<details>
<summary>Expand output</summary>

  ● Test suite failed to run

    Cannot find module 'next/server/web/spec-extension/response' from 'unit/web-runtime/next-response.test.ts'

       5 | import { NextResponse } from 'next/server/web/spec-extension/response'
       6 |
    >  7 | const toJSON = async (response: Response) => ({
         |                   ^
       8 |   body: await response.json(),
       9 |   contentType: response.headers.get('content-type'),
      10 |   status: response.status,

      at Resolver.resolveModule (../node_modules/.pnpm/jest-resolve@27.5.1/node_modules/jest-resolve/build/resolver.js:324:11)
      at Object.<anonymous> (unit/web-runtime/next-response.test.ts:7:19)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/src-folder | Change |
| - | - | - | - |
| buildDuration | 34.8s | 35.1s | ⚠️ +276ms |
| buildDurationCached | 14.5s | 10.2s | -4.3s |
| nodeModulesSize | 95.1 MB | 95.1 MB | ✓ |
| nextStartRea..uration (ms) | 276ms | 274ms | -2ms |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/src-folder | Change |
| - | - | - | - |
| 152-HASH.js gzip | 64.5 kB | 64.5 kB | ✓ |
| 53.HASH.js gzip | 182 B | 182 B | ✓ |
| main-app-HASH.js gzip | 205 B | 205 B | ✓ |
| main-HASH.js gzip | 79.2 kB | 79.2 kB | ✓ |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 146 kB | 146 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/src-folder | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/src-folder | Change |
| - | - | - | - |
| _app-HASH.js gzip | 192 B | 192 B | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 484 B | 484 B | ✓ |
| css-HASH.js gzip | 804 B | 804 B | ✓ |
| dynamic-HASH.js gzip | 2.29 kB | 2.29 kB | ✓ |
| edge-ssr-HASH.js gzip | 255 B | 255 B | ✓ |
| head-HASH.js gzip | 827 B | 827 B | ✓ |
| hooks-HASH.js gzip | 848 B | 848 B | ✓ |
| image-HASH.js gzip | 4.3 kB | 4.3 kB | ✓ |
| index-HASH.js gzip | 252 B | 252 B | ✓ |
| link-HASH.js gzip | 2.69 kB | 2.69 kB | ✓ |
| routerDirect..HASH.js gzip | 782 B | 782 B | ✓ |
| script-HASH.js gzip | 857 B | 857 B | ✓ |
| withRouter-HASH.js gzip | 781 B | 781 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.6 kB | 15.6 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/src-folder | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 481 B | ✓ |
| Overall change | 481 B | 481 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/src-folder | Change |
| - | - | - | - |
| index.html gzip | 489 B | 489 B | ✓ |
| link.html gzip | 504 B | 504 B | ✓ |
| withRouter.html gzip | 485 B | 485 B | ✓ |
| Overall change | 1.48 kB | 1.48 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/src-folder | Change |
| - | - | - | - |
| edge-ssr.js gzip | 110 kB | 110 kB | ✓ |
| page.js gzip | 98.4 kB | 98.4 kB | ✓ |
| Overall change | 209 kB | 209 kB | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/src-folder | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 584 B | 585 B | ⚠️ +1 B |
| middleware-r..fest.js gzip | 145 B | 145 B | ✓ |
| middleware.js gzip | 27 kB | 27 kB | ✓ |
| edge-runtime..pack.js gzip | 1.83 kB | 1.83 kB | ✓ |
| Overall change | 29.6 kB | 29.6 kB | ⚠️ +1 B |

</details>

#### Diffs
<details>
<summary>Diff for <strong>page.js</strong></summary>

```diff
@@ -8618,23 +8618,23 @@ Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime`;
       /***/
     },
 
-    /***/ 1036: /***/ (
+    /***/ 1028: /***/ (
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
     ) => {
-      Promise.resolve(/* import() eager */).then(
-        __webpack_require__.bind(__webpack_require__, 1210)
-      );
       Promise.resolve(/* import() eager */).then(
         __webpack_require__.bind(__webpack_require__, 2973)
       );
       Promise.resolve(/* import() eager */).then(
-        __webpack_require__.bind(__webpack_require__, 6448)
+        __webpack_require__.bind(__webpack_require__, 1210)
       );
       Promise.resolve(/* import() eager */).then(
         __webpack_require__.bind(__webpack_require__, 9916)
       );
+      Promise.resolve(/* import() eager */).then(
+        __webpack_require__.bind(__webpack_require__, 6448)
+      );
 
       /***/
     },
```
</details>
<details>
<summary>Diff for <strong>middleware-b..-manifest.js</strong></summary>

```diff
@@ -9,7 +9,7 @@ self.__BUILD_MANIFEST = {
   rootMainFiles: [
     "static/chunks/webpack-c452a3e31b73f504.js",
     "static/chunks/152-4025b906f33b8271.js",
-    "static/chunks/main-app-c9a0aaff35eafcb5.js"
+    "static/chunks/main-app-ef0948581cabd144.js"
   ],
   pages: {
     "/": [
```
</details>
<details>
<summary>Diff for <strong>main-app-HASH.js</strong></summary>

```diff
@@ -1,23 +1,23 @@
 (self["webpackChunk_N_E"] = self["webpackChunk_N_E"] || []).push([
   [744],
   {
-    /***/ 9915: /***/ function(
+    /***/ 5764: /***/ function(
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
     ) {
-      Promise.resolve(/* import() eager */).then(
-        __webpack_require__.t.bind(__webpack_require__, 429, 23)
-      );
       Promise.resolve(/* import() eager */).then(
         __webpack_require__.t.bind(__webpack_require__, 1161, 23)
       );
       Promise.resolve(/* import() eager */).then(
-        __webpack_require__.t.bind(__webpack_require__, 8138, 23)
+        __webpack_require__.t.bind(__webpack_require__, 429, 23)
       );
       Promise.resolve(/* import() eager */).then(
         __webpack_require__.t.bind(__webpack_require__, 9379, 23)
       );
+      Promise.resolve(/* import() eager */).then(
+        __webpack_require__.t.bind(__webpack_require__, 8138, 23)
+      );
 
       /***/
     }
@@ -28,7 +28,7 @@
       return __webpack_require__((__webpack_require__.s = moduleId));
     };
     /******/ __webpack_require__.O(0, [152], function() {
-      return __webpack_exec__(7070), __webpack_exec__(9915);
+      return __webpack_exec__(7070), __webpack_exec__(5764);
     });
     /******/ var __webpack_exports__ = __webpack_require__.O();
     /******/ _N_E = __webpack_exports__;
```
</details>
</details>
Commit: adbe2951dd2024dcf67a59b7aea29db76a4304d0

File: .prettierignore [hunk 1]
```diff
@@ -2,9 +2,9 @@ node_modules
 **/.next/**
 **/_next/**
 **/dist/**
-packages/next/bundles/webpack/packages/*.runtime.js
-packages/next/bundles/webpack/packages/lazy-compilation-*.js
-packages/next/compiled/**
+packages/next/src/bundles/webpack/packages/*.runtime.js
+packages/next/src/bundles/webpack/packages/lazy-compilation-*.js
+packages/next/src/compiled/**
```
Comment (SukkaW) [inline .prettierignore:7]: IMHO `packages/next/compiled/` should not be moved into `src` folder.

Comment (timneutkens) [reply .prettierignore:7]: This is intentional, it's the checked-in source code, not the part that is published.

**Review by hanneslund — APPROVED** [review]

**Pull request merged**